### PR TITLE
Transform PDF

### LINKED ISSUE #74: Report: Transform waypoint data into PDF report
### Task

- [x] Transform waypoint data to generate performance statistics
- [x] Create report to include statistics.

### User Story
As a developer, I want to transform the waypoint data so that I can generate useful statistics that can be sent out to a user.

### PR DESCRIPTION
**Pull request by A-Faris**
Create PDF summary report
closes #74

### CODE REVIEW

File: pdf_report/transform_pdf.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
+"""Create PDF with train summary details"""
+
+from psycopg2.extensions import connection
+
+from extract_pdf import get_connection, query_db
+
+# [RealDictRow({'station_id': 7, 'waypoint_id': 8407, 'run_date': datetime.date(2024, 7, 22), 'booked_arrival': datetime.datetime(2024, 7, 22, 23, 35), 'actual_arrival': datetime.datetime(2024, 7, 22, 23, 34), 'booked_departure': datetime.datetime(2024, 7, 22, 23, 39), 'actual_departure': datetime.datetime(2024, 7, 22, 0, 1), 'service_id': 5097, 'station_crs': 'YRK', 'station_name': 'York Rail Station'})]
```
Comment (fm1psy) [inline pdf_report/transform_pdf.py]: Was this comment a personal note? perhaps this note is no longer needed.
```suggestion
```

Comment (FarihaChoudhury) [reply pdf_report/transform_pdf.py]: ew why are there comments in a pull request? fix up

File: pdf_report/extract_pdf.py [hunk 2]
```diff
@@ -28,16 +28,21 @@ def get_cursor(conn: connection) -> cursor:
     return conn.cursor(cursor_factory=RealDictCursor)
 
 
+def query_db(conn: connection, query: str) -> list[tuple]:
```
Comment (fm1psy) [inline pdf_report/extract_pdf.py:31]: I like the use of a function to run queries, helps reduce dry code

File: .util/pylint_scores.txt [hunk 3]
```diff
@@ -29,6 +29,14 @@ Your code has been rated at 10.00/10
 ------------------------------------
 Your code has been rated at 10.00/10
 
+************* Module transform_pdf
+pdf_report/transform_pdf.py:40:0: C0301: Line too long (122/100) (line-too-long)
+pdf_report/transform_pdf.py:41:0: C0301: Line too long (127/100) (line-too-long)
+pdf_report/transform_pdf.py:76:0: C0116: Missing function or method docstring (missing-function-docstring)
+
+-----------------------------------
+Your code has been rated at 8.93/10
```
Comment (Zhi-704) [inline .util/pylint_scores.txt]: Smh has to be above 9

**Review by fm1psy — DISMISSED** [review]
This looks ready to merge

File: pdf_report/transform_pdf.py (hunk 1 above)
Comment (A-Faris) [reply pdf_report/transform_pdf.py]: removed it

File: pdf_report/extract_pdf.py (hunk 2 above)
Comment (A-Faris) [reply pdf_report/extract_pdf.py:31]: thank you

File: pdf_report/transform_pdf.py [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+
+
+def convert_html_to_pdf(source_html: str, output_filename: str) -> bool:
+    """Outputs HTML to a target file."""
+    with open(output_filename, "w+b") as f:
+        pisa_status = pisa.CreatePDF(source_html, dest=f)
+    return pisa_status.err
+
+
+def transform_pdf() -> None:
```
Comment (FarihaChoudhury) [inline pdf_report/transform_pdf.py]: I think this function as a whole is quite big, maybe you could separate it into smaller functions by similarity

File: pdf_report/transform_pdf.py [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+    station_info = delayed_df[['station_name',
+                               'station_crs']].drop_duplicates()
+
+    station_info_html = "<p><b>Station CRS</b> - <b>Station Name</b></p><ul>"
+    for _, row in station_info.iterrows():
+        station_info_html += f"<li>{row['station_crs']}\
+              - {row['station_name']}</li>"
+    station_info_html += "</ul>"
+
+    html_report = f"""
```
Comment (FarihaChoudhury) [inline pdf_report/transform_pdf.py]: could have a function that just returns this html report? to reduce the clutter in ```transform_pdf.py```

Comment (A-Faris) [reply pdf_report/transform_pdf.py]: Done

File: .util/pylint_scores.txt (hunk 3 above)
Comment (A-Faris) [reply .util/pylint_scores.txt]: Done

File: pdf_report/transform_pdf.py [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+        avg_delay_df, 'station_crs',
+        ['avg_arrive_delay_minutes',
+            'avg_departure_delay_minutes'], 'Average Delay by Station',
+        'Average Delay (minutes)')
+    avg_delay_long_chart = generate_grouped_bar_chart(
+        avg_delay_long_df, 'station_crs',
+        ['avg_arrive_delay_long_minutes', 'avg_departure_delay_long_minutes'],
+        'Average Delay more than 1 min by Station', 'Average Delay more than 1 min (minutes)')
+
+    html_report = generate_html_report(delayed_df, cancellation_chart,
```
Comment (FarihaChoudhury) [inline pdf_report/transform_pdf.py:257]: looks so much better

**Review by FarihaChoudhury — DISMISSED** [review]
Good work :) love the report

**Review by FarihaChoudhury — APPROVED** [review]
amazin

**Pull request merged**